optuna/search_space/intersection.py

```diff
@@ -26,12 +26,18 @@ def _calculate(
     if include_pruned:
         states_of_interest.append(optuna.trial.TrialState.PRUNED)
 
-    trials_of_interest = [trial for trial in trials if trial.state in states_of_interest]
+    next_cached_trial_number = -1
+
+    found_trial_of_interest = False
+
+    for trial in reversed(trials):
+        if trial.state not in states_of_interest:
+            continue
+
+        elif not found_trial_of_interest:
```

Comment:
How about checking `next_cached_trial_number` is equal to `-1`? It will be simpler than using `found_trial_of_interest`.

```suggestion
        if next_cached_trial_number == -1:
```

Reply:
Using a Boolean flag is faster especially when repeatedly applied in a loop but reduces readability?